tika-pipes/tika-fetchers/tika-fetcher-microsoft-graph/src/main/java/org/apache/tika/pipes/fetchers/microsoftgraph/MicrosoftGraphFetcher.java

```diff
@@ -0,0 +1,140 @@
+
+    public MicrosoftGraphFetcher() {
+
+    }
+
+    public MicrosoftGraphFetcher(MsGraphFetcherConfig msGraphFetcherConfig) {
+        this.msGraphFetcherConfig = msGraphFetcherConfig;
+    }
+
+    /**
+     * Set seconds to throttle retries as a comma-delimited list, e.g.: 30,60,120,600
+     *
+     * @param commaDelimitedLongs
+     * @throws TikaConfigException
+     */
+    @Field
+    public void setThrottleSeconds(String commaDelimitedLongs) throws TikaConfigException {
+        String[] longStrings = commaDelimitedLongs.split(",");
+        long[] seconds = new long[longStrings.length];
+        for (int i = 0; i < longStrings.length; i++) {
+            try {
+                seconds[i] = Long.parseLong(longStrings[i]);
+            } catch (NumberFormatException e) {
+                throw new TikaConfigException(e.getMessage());
+            }
+        }
+        setThrottleSeconds(seconds);
+    }
+
+    public void setThrottleSeconds(long[] throttleSeconds) {
+        this.throttleSeconds = throttleSeconds;
+    }
+
+    @Override
+    public void initialize(Map<String, Param> map) {
+        String[] scopes = msGraphFetcherConfig.getScopes().toArray(new String[0]);
+        if (msGraphFetcherConfig.getCredentials() instanceof ClientCertificateCredentialsConfig) {
+            ClientCertificateCredentialsConfig credentials =
+                    (ClientCertificateCredentialsConfig) msGraphFetcherConfig.getCredentials();
+            graphClient = new GraphServiceClient(
+                    new ClientCertificateCredentialBuilder().clientId(credentials.getClientId())
+                            .tenantId(credentials.getTenantId()).pfxCertificate(
+                                    new ByteArrayInputStream(credentials.getCertificateBytes()))
+                            .clientCertificatePassword(credentials.getCertificatePassword())
+                            .build(), scopes);
+        } else if (msGraphFetcherConfig.getCredentials() instanceof ClientSecretCredentialsConfig) {
+            ClientSecretCredentialsConfig credentials =
+                    (ClientSecretCredentialsConfig) msGraphFetcherConfig.getCredentials();
+            graphClient = new GraphServiceClient(
+                    new ClientSecretCredentialBuilder().tenantId(credentials.getTenantId())
+                            .clientId(credentials.getClientId())
+                            .clientSecret(credentials.getClientSecret()).build(), scopes);
+        }
+    }
+
+    @Override
+    public void checkInitialization(InitializableProblemHandler initializableProblemHandler)
+            throws TikaConfigException {
+    }
+
+    @Override
+    public InputStream fetch(String fetchKey, Metadata metadata) throws TikaException, IOException {
+        int tries = 0;
+        Exception ex;
+        do {
+            try {
+                long start = System.currentTimeMillis();
+                String[] fetchKeySplit = fetchKey.split(",");
```

Comment:
Download URLs provided for each drive item seem to be unique (we can confirm this) and would prevent this specific nature of a fetch key being provided. This would simplify the call to the fetcher as well.

Thoughts?

Reply:
I'm stepping back on this comment. After reviewing the responses from the MS Graph API, there is no simple (without creating additional resources in the API) to get a long-living download URL. The URLs it provides are short lived (typically one hour) with a temp auth token. This is not ideal for Tika

The drive item ID approach here I believe is the right one. I'll be testing it soon.